### PR TITLE
Annotations for Guava as specified in Stubs used at Google

### PR DESCRIPTION
**Pull request by dilraj45**
Incorporated the annotations as specified in stubs files for Guava being used at Google

### CODE REVIEW

File: guava/src/com/google/common/base/Optional.java [hunk 1]
```diff
@@ -121,7 +124,7 @@
    * @since 21.0
    */
   @NullableDecl
-  public static <T> Optional<T> fromJavaUtil(@NullableDecl java.util.Optional<T> javaUtilOptional) {
+  public static <T extends Object> @PolyNull Optional<T> fromJavaUtil(@NullableDecl java.util.Optional<T> javaUtilOptional) {
```
Comment (mernst) [inline guava/src/com/google/common/base/Optional.java]: It usually doesn't make sense to write just one occurrence of `@PolyNull`.
See https://checkerframework.org/manual/#qualifier-polymorphism-multiple-qualifiers and https://checkerframework.org/manual/#qualifier-polymorphism-single-qualifier .
I suspect you will want to change the annotations on this method and the next one.

File: guava/src/com/google/common/base/Joiner.java [hunk 2]
```diff
@@ -40,7 +40,6 @@
  *  . . .
  * return joiner.join("Harry", null, "Ron", "Hermione");
  * }</pre>
- *
```
[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Joiner.java]: Okay, I'll make sure to avoid whitespace changes

File: guava/src/com/google/common/base/Optional.java [hunk 3]
```diff
@@ -80,14 +83,14 @@
  * @since 10.0
  */
 @GwtCompatible(serializable = true)
-public abstract class Optional<T> implements Serializable {
+public abstract class Optional<T extends Object> implements Serializable {
```
[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: These changes are mainly for restrict the instantiation of generic classes to only `@NonNull` types, [see](https://checkerframework.org/manual/#generics) or to default type annotation to be more precise, [see](https://checkerframework.org/manual/#faq-semantics-section). Please let me now if my understanding is at fault. Thanks!

File: guava/src/com/google/common/base/Optional.java (hunk 1 above)
Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: I'll make the changes as suggested

File: guava/src/com/google/common/base/Optional.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/base/Optional.java]: This makes sense, thanks.

[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: In stub file shared by you, developers have added the following note regarding usage of `extends Object`
```
// NOTE: "extends Object" means there's no explicit annotation so none is added.
// Which means that the default interpretation is used. Since the default
// changed between checker framework versions, "extends Object" translates to:
// - "extends @NonNull Object" in v2.0
// - "extends @Nullable Object" in v2.1+
// TODO: Remove "extends Object".
```
I think I should revert these changes and further should use the fully annotated versions like `extends @NonNull Object` or `extends @Nullable Object` instead of relying on the default interpretation. Let me know how should I proceed. Thanks!

[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: Initially I was read through the stub and developed a rationale for including `extends Object`. But later on when I carefully examined the stub file I noticed these comments, which made me question whether or not to include this.
Should I remove these (as developers wanted) or keep these changes?
Thank you.

[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/base/Optional.java]: I think smaller diffs are desirable, so I wouldn't add `extends Object` except where necessary.

File: guava/src/com/google/common/collect/Iterables.java [hunk 4]
```diff
@@ -786,8 +783,7 @@ public void forEach(Consumer<? super T> action) {
    * @return the first element of {@code iterable} or the default value
    * @since 7.0
    */
-  @NullableDecl
-  public static <T> T getFirst(Iterable<? extends T> iterable, @NullableDecl T defaultValue) {
+  public static <T extends @Nullable Object> @PolyNull T getFirst(Iterable<? extends T> iterable, @PolyNull T defaultValue) {
```
[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/collect/Iterables.java]: I don't understand this comment.

What does "as per the stub file" mean?  If it's in the stub file, and this pull request is for the stub file, then why isn't it already in the diffs?

Why do you think `@Nullable` will be more appropriate?  Please explain your reasoning, or I cannot comment on it.

File: guava/src/com/google/common/util/concurrent/ListenableFutureTask.java [hunk 5]
```diff
@@ -34,7 +34,7 @@
  * @since 1.0
  */
 @GwtIncompatible
-public class ListenableFutureTask<V> extends FutureTask<V> implements ListenableFuture<V> {
+public class ListenableFutureTask<V extends @Nullable Object> extends FutureTask<V> implements ListenableFuture<V> {
```
[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/util/concurrent/ListenableFutureTask.java]: The code contains three unrelated type variables (all named `V`).  The first `<V>' in ```static <V> ListenableFutureTask<V> create(...)``` declares a new type variable.  All three of the type variables should have consistent declarations.

File: guava/src/com/google/common/collect/Iterables.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/collect/Iterables.java]: By "as per the stub file", I meant the annotations that are specified in the stub file for the given class. 

I have added the annotations as you can see in the diff.

I was just thinking if this could be replaced with @Nullable annotation, for this method is just returning value from another method which is: `public static <T> @Nullable T getNext(Iterator<? extends T> iterator, @Nullable T defaultValue)`. And return type of that method is annotated to be @Nullable, so I thought it might be better annotated as @Nullable. But I haven't made this change, I just wanted to have your opinion on this. Thanks.

File: guava/src/com/google/common/base/Optional.java [hunk 6]
```diff
@@ -282,7 +284,7 @@
    * @throws NullPointerException if the function returns {@code null}
    * @since 12.0
    */
-  public abstract <V> Optional<V> transform(Function<? super T, V> function);
+  public abstract <V extends @NonNull Object> Optional<V> transform(Function<? super T, V> function);
```
Comment (mernst) [inline guava/src/com/google/common/base/Optional.java]: I think that every occurrence of `extends @NonNull Object` in the diffs is unnecessary.  I suggest removing them.

File: guava/src/com/google/common/collect/ImmutableList.java [hunk 7]
```diff
@@ -166,7 +167,7 @@
    *
    * @throws NullPointerException if any element is null
    */
-  public static <E> ImmutableList<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7, E e8, E e9) {
+  public static <E extends @NonNull Object> ImmutableList<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7, E e8, E e9) {
```
Comment (mernst) [inline guava/src/com/google/common/collect/ImmutableList.java]: The Google Java style requires lines of code to break before 100 characters.  See how the next `of` method was formatted, and do so for the lines that are longer than 100 characters.

(But, in this case `extends @NonNull Object` seems unnecessary.)

File: guava/src/com/google/common/base/Verify.java [hunk 8]
```diff
@@ -473,7 +475,8 @@ public static void verify(
    * @see Preconditions#checkNotNull Preconditions.checkNotNull()
    */
   @CanIgnoreReturnValue
-  public static <T> T verifyNotNull(@Nullable T reference) {
+  @EnsuresNonNull("#1")
+  public static <T extends @Nullable Object> @NonNull T verifyNotNull(@Nullable T reference) {
```
Comment (mernst) [inline guava/src/com/google/common/base/Verify.java]: verifyNotNull should only be called on a non-null reference.  You can just remove all three annotations from this line.

File: guava/src/com/google/common/collect/ImmutableCollection.java [hunk 9]
```diff
@@ -163,7 +163,7 @@
 @SuppressWarnings("serial") // we're overriding default serialization
 // TODO(kevinb): I think we should push everything down to "BaseImmutableCollection" or something,
 // just to do everything we can to emphasize the "practically an interface" nature of this class.
-public abstract class ImmutableCollection<E> extends AbstractCollection<E> implements Serializable {
+public abstract class ImmutableCollection<E extends @Nullable Object> extends AbstractCollection<E> implements Serializable {
```
Comment (mernst) [inline guava/src/com/google/common/collect/ImmutableCollection.java]: The Google stubs have `extends @NonNull Object` here, and the class Javadoc says the class will never contain a null element.

File: guava/src/com/google/common/base/Preconditions.java [hunk 10]
```diff
@@ -875,7 +903,9 @@ public static void checkState(
    * @see Verify#verifyNotNull Verify.verifyNotNull()
    */
   @CanIgnoreReturnValue
-  public static <T> T checkNotNull(T reference) {
+  @EnsuresNonNull("#1")
+  @Pure
+  public static <T extends @Nullable Object> @NonNull T checkNotNull(T reference) {
```
Comment (mernst) [inline guava/src/com/google/common/base/Preconditions.java]: checkNotNull should never be called on a null value.  You can remove all the annotations on this line, leaving it as it was before your changes.  The two added declaration annotations are good, though.

File: guava/src/com/google/common/collect/Lists.java [hunk 11]
```diff
@@ -527,7 +529,8 @@ public E get(int index) {
    * java.util.stream.Stream#map}. This method is not being deprecated, but we gently encourage you
    * to migrate to streams.
    */
-  public static <F, T> List<T> transform(
+  @DefaultQualifier(value = Nullable.class, locations = TypeUseLocation.UPPER_BOUND)
```
Comment (mernst) [inline guava/src/com/google/common/collect/Lists.java]: Don't use this just to save a couple annoattions; write the couple annotations explicitly, if indeed any are needed.

File: guava/src/com/google/common/collect/Lists.java [hunk 12]
```diff
@@ -84,7 +86,7 @@ private Lists() {}
    * directly, taking advantage of the new <a href="http://goo.gl/iz2Wi">"diamond" syntax</a>.
    */
   @GwtCompatible(serializable = true)
-  public static <E> ArrayList<E> newArrayList() {
+  public static <E extends @Nullable Object> ArrayList<E> newArrayList() {
```
Comment (mernst) [inline guava/src/com/google/common/collect/Lists.java]: Make newArrayListWithExpectedSize consistent with newArrayList.

File: guava/src/com/google/common/base/Optional.java (hunk 6 above)
Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: Explicit and Implicit bounds for generics are treated differently by the Checker Framework. Here is a link to documentation that explains the difference, https://checkerframework.org/manual/#defaults-classfile.

By writing `V extends @NonNull Object` we are restricting the instantiation to `@NonNull` types which would have otherwise defaulted to type `@Nullable` as per the Checker Framework's defaulting rules. For more information on restricting the instantiation of generic class please the refer to the link: https://checkerframework.org/manual/#generics

If the usage `extends @NonNull Object` is logically incorrect then please let me know, so that I can remove these annotations in that case. Further, I think every occurrence of `extends @Nullable Object` is redundant and can be removed.

Thanks!

File: guava/src/com/google/common/collect/Lists.java (hunk 11 above)
Comment (dilraj45) [reply guava/src/com/google/common/collect/Lists.java]: In this particular case, I think the annotations are already in place in the diff and just removing the `@DefaultQualifier` annotation will suffice.

Comment (mernst) [reply guava/src/com/google/common/collect/Lists.java]: That is correct.

File: guava/src/com/google/common/base/Optional.java (hunk 6 above)
Comment (mernst) [reply guava/src/com/google/common/base/Optional.java]: > Explicit and Implicit bounds for generics are treated differently by the Checker Framework. Here is a link to documentation that explains the difference, https://checkerframework.org/manual/#faq-semantics-section.

You gave a link to a section about unannotated libraries.  This is not relevant to your work.

To support your claim, it would be helpful for you to give a concrete example of code in which explicit and implicit bounds are treated differently by the Checker Framework.

In particular, you coud write code like this:

```
   <V extends @NonNull Object> ReturnType methodName(args...) { ... }
   <V extends @Nullable Object> ReturnType methodName(args...) { ... }
   <V extends Object> ReturnType methodName(args...) { ... }
```

where everything after the type argument declaration (that is, everything starting with `ReturnType`) is identical, but the Checker Framework treats all three methods differently when type-checking clients.

Comment (dilraj45) [reply guava/src/com/google/common/base/Optional.java]: Please consider the following given example
```java
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

public class Test {
  public static <V extends @NonNull Object> String dummyA(V argA) {
    return argA.toString();
  }
  
  public static <V extends @Nullable Object> String dummyB(V argB) {
    return argB.toString();
  }

  public static <V extends Object> String dummyC(V argC) {
    return argC.toString();
  }

  public static <V> String dummyD(V argD) {
    return argD.toString();
  }

  public static <V extends @Nullable Object> V dummyE(V argE) {
    return null;
  }

  public static <@Nullable V extends @Nullable Object> V dummyF(V argF) {
    return null;
  }

  public static void main(String[] args) {
    @Nullable String nullableStr = null;
    String nonnullStr = "Dummy";
    
    dummyA(nullableStr);
    dummyB(nullableStr);
    dummyC(nullableStr);
    dummyD(nullableStr);

    dummyA(nonnullStr);
    dummyB(nonnullStr);
    dummyC(nonnullStr);
    dummyD(nonnullStr);
  }
}
```
Following depicts the compile-time errors thrown by nullness type processor
```
Test.java:10: error: [dereference.of.nullable] dereference of possibly-null reference argB
    return argB.toString();
           ^
Test.java:18: error: [dereference.of.nullable] dereference of possibly-null reference argD
    return argD.toString();
           ^
Test.java:22: error: [return.type.incompatible] incompatible types in return.
    return null;
           ^
  found   : @FBCBottom @Nullable  null
  required: V[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
Test.java:33: error: [type.argument.type.incompatible] incompatible types in type argument.
    dummyA(nullableStr);
          ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull Object
Test.java:35: error: [type.argument.type.incompatible] incompatible types in type argument.
    dummyC(nullableStr);
          ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull Object
5 errors
```
Conclusion
* `@NonNull` is treated as the implicit lower bound. Evident from the error message for dummyE which returns null as a value.
* Both `V extends @NonNull Object` and `V extends Object` are equivalent and are treated as `V [extends @NonNull Object super @NonNull Void]` , these both restrict the type V to `@NonNull` types only. Thus passing a `@Nullable` type reference causes checker to throw an error, as evident from the error message for method dummyA and dummyC.
* Both `V` and `V extends @Nullable Object` are identical and are treated as `V [extends @Nullable Object super @NonNull Void]`

This is in accordance with the defaulting rules documented in the manual, see, https://checkerframework.org/manual/#climb-to-top, which says the following about implicit upper and lower bounds
>Implicit upper bounds are defaulted to top to allow them to be instantiated in any way. If a user declared class C\<T\> { ... }, then we assume that the user intended to allow any instantiation of the class, and the declaration is interpreted as class C\<T extends `@Nullable` Object\> { ... } rather than as class C\<T extends `@NonNull` Object\> { ... }.

>Implicit lower bounds are defaulted to the bottom type, again to allow maximal instantiation. Note that Java does not allow a programmer to express both the upper and lower bounds of a type, but the Checker Framework allows the programmer to specify either or both

Thanks!

**Comment by dilraj45** [comment]
@mernst I think I have addressed your comments. Can you please have another look? Thanks!

**Comment by mernst** [comment]
Can you pull the master branch of typetools/guava into your branches?

File: guava/src/com/google/common/collect/Lists.java [hunk 13]
```diff
@@ -53,6 +53,8 @@
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.dataflow.qual.Pure;
 import org.checkerframework.framework.qual.AnnotatedFor;
+import org.checkerframework.framework.qual.DefaultQualifier;
```
Comment (mernst) [inline guava/src/com/google/common/collect/Lists.java]: You don't need these two import statements.

File: guava/src/com/google/common/collect/Ordering.java [hunk 14]
```diff
@@ -491,7 +491,7 @@ protected Ordering() {}
    * @param comparators the comparators to try in order
    */
   @GwtCompatible(serializable = true)
-  public static <T> Ordering<T> compound(Iterable<? extends Comparator<? super T>> comparators) {
+  public static <T extends @Nullable Object > Ordering<T> compound(Iterable<? extends Comparator<? super T>> comparators) {
```
Comment (mernst) [inline guava/src/com/google/common/collect/Ordering.java]: Stray space within `Object >`.  But, you don't need this change since it's the default.
Omitting changes like this one will keep the diffs smaller.
You may also be able to remove some `import ...Nullable;` statements as a result, and avoid some line breaks.

**Comment by mernst** [comment]
Every `Immutable*` class in Guava is null-hostile:  No null value should every appear in it.

The Google stub files handle a subset of those.  Furthermore, it's neither complete nor consistent.  For example, the Google stub files contain annotations for ImmutableList.of but not ImmutableMap.of.

I propose that we make this pull request be largely consistent with the Google stubs, so we can get it merged.  In other words, there are not changes that need to be made to this pull request.

However, can you please make another pull request that handles only null-hostility of all `Immutable*` classes, and nothing more?  (It won't necessarily be a complete annotation of those classes.)  I think making those all at once will make that particular pull request easier to perform and check, and will make us less likely to suffer oversights while handling null-hostility.

File: guava/src/com/google/common/base/Preconditions.java [hunk 15]
```diff
@@ -875,7 +903,9 @@ public static void checkState(
    * @see Verify#verifyNotNull Verify.verifyNotNull()
    */
   @CanIgnoreReturnValue
-  public static <T> T checkNotNull(T reference) {
+  @EnsuresNonNull("#1")
+  @Pure
+  public static <T> @NonNull T checkNotNull(@NonNull T reference) {
```
[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Preconditions.java]: Is `@EnsuresNonNull` necessary here?

[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/base/Preconditions.java]: No.  The routine should only be called on a non-null value, so the `@EnsuresNonNull` would have no effect.

**Comment by mernst** [comment]
Please don't force-push, rebase, or otherwise change the history of your branch.  (It makes life more difficult for people who have a clone of your repository.)  Instead, just add new commits to the history.

File: guava/src/com/google/common/base/Preconditions.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (mernst) [reply guava/src/com/google/common/base/Preconditions.java]: (I already gave this comment on `Verify.verifyNotNull`.)

[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Preconditions.java]: Thanks! I just wanted to confirm checkNotNull as well.

File: guava/src/com/google/common/util/concurrent/AbstractCheckedFuture.java [hunk 16]
```diff
@@ -21,6 +21,7 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (mernst) [inline guava/src/com/google/common/util/concurrent/AbstractCheckedFuture.java]: The import statement is not needed here.

File: guava/src/com/google/common/util/concurrent/AsyncCallable.java [hunk 17]
```diff
@@ -17,6 +17,7 @@
 import com.google.common.annotations.Beta;
 import com.google.common.annotations.GwtCompatible;
 import java.util.concurrent.Future;
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (mernst) [inline guava/src/com/google/common/util/concurrent/AsyncCallable.java]: The import statement is not needed here.

File: guava/src/com/google/common/util/concurrent/CheckedFuture.java [hunk 18]
```diff
@@ -22,6 +22,7 @@
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (mernst) [inline guava/src/com/google/common/util/concurrent/CheckedFuture.java]: The import statement is not needed here.  (Check elsewhere too.)

File: guava/src/com/google/common/util/concurrent/Futures.java [hunk 19]
```diff
@@ -189,7 +189,7 @@ private Futures() {}
   // TODO(b/72241893): Remove by 2018-07
   @Deprecated
   @GwtIncompatible // TODO
-  public static <V, X extends Exception> CheckedFuture<V, X> immediateCheckedFuture(
+  public static <V, X extends Exception> CheckedFuture<V, X> immediateCheckedFuture (
```
Comment (mernst) [inline guava/src/com/google/common/util/concurrent/Futures.java]: Don't make unnecessary whitespace changes.

File: guava/src/com/google/common/base/Optional.java [hunk 20]
```diff
@@ -282,7 +284,7 @@
    * @throws NullPointerException if the function returns {@code null}
    * @since 12.0
    */
-  public abstract <V> Optional<V> transform(Function<? super T, V> function);
+  public abstract <V extends Object> Optional<V> transform(Function<? super T, V> function);
```
Comment (mernst) [inline guava/src/com/google/common/base/Optional.java]: Your version is equivalent to `<V extends @NonNull Object>`, and is therefore correct.
However, I suggest that you include the qualifier, for emphasis.  Otherwise, someone who is reading the code might think that the `extends Object` is irrelevant (which it is in plain Java), and some IDEs will suggest removing it.

**Comment by dilraj45** [comment]
@mernst, I have addressed your comments, please have an another look

File: guava/src/com/google/common/base/Preconditions.java [hunk 21]
```diff
@@ -916,8 +946,9 @@ public static void checkState(
    * @see Verify#verifyNotNull Verify.verifyNotNull()
    */
   @CanIgnoreReturnValue
-  public static <T> T checkNotNull(
-      T reference, @Nullable String errorMessageTemplate, Object @Nullable... errorMessageArgs) {
+  @Pure
+  public static <T extends @NonNull Object> T checkNotNull(T reference,
```
[the remark this answers is not in the document]
Comment (dilraj45) [reply guava/src/com/google/common/base/Preconditions.java]: I was under the impression that this conforms to Google's formatting style. I did it to make it consistent with formatting for other `checkNotNull` methods. I'll fix this and the formatting for those methods as well. Thanks!

**Comment by dilraj45** [comment]
I have once again gone through all the stubs files, so as to ensure that I'm not missing on anything. @mernst Please have another look.

**Comment by dilraj45** [comment]
I have created a pull request for annotating `ImmutableXXX` classes. Here is the link for the same, https://github.com/typetools/guava/pull/17. I have already added changes for a few classes and will be adding changes for the remaining classes as well. I was thinking, If you can please review the pull requested changes while I work on the remaining classes, that will be really helpful. Thanks!

**Comment by mernst** [comment]
Thanks.  I made some final changes, and this looks good now.
